### DISCUSSION TITLE
SciSpaCy vs SpaCy Entity Linker

### DISCUSSION THREAD

**Discussion by amirj** [explosion/spaCy#8130] [Help: Coding & Implementations] [answered]
I have both SciSpaCy and SpaCy entity linker in my pipeline. I'm going to leverage SciSpaCy for candidate generation and then leverage SpaCy for disambiguation. When I'm using SpaCy for disambiguation, I always get a lower P/R values in comparison to the baseline SciSpaCy. This is my custom candidate generation function:

```python
@registry.misc("scispacy_candidate_generator")
def create_candidates():
    """Custom candidate generation function"""
    return get_scispacy_candidates


def get_scispacy_candidates(kb, span):
    """Leverage SciSpaCy to generate a list of plausible candidates"""
    candidates = []
    # collect plausible candidates from scispacy
    for cui, score in span._.kb_ents:
        candidate = Candidate(
            kb=kb,
            entity_hash=kb.vocab.strings[cui],
            entity_freq=111,
            entity_vector=kb.get_vector(cui),
            alias_hash=kb.vocab.strings[span.text],
            prior_prob=float(score),
        )
        candidates.append(candidate)
    return candidates
```

As a sanity check, I set SciSpaCy to generate **only one candidate**, i.e. the above function returns a list that contains only one candidate per input span. With this configuration, I'm getting the same result as the baseline SciSpaCy which is expected. But when I increase the number of candidates to greater than one (e.g. 5), I'm getting lower P/R values in comparison to the baseline SciSpaCy candidate generation:

```python
config = {
                "labels_discard": [],
                "n_sents": 2,
                "incl_prior": True,
                "incl_context": False,
                "model": DEFAULT_NEL_MODEL,
                "entity_vector_length": self.entity_vector_length,
                "get_candidates": {"@misc": "scispacy_candidate_generator"},
            }
```

I'm expecting to get the same result when having multiple candidates from SciSpaCy with `"incl_prior": True` and `"incl_context": False`.

**Comment by amirj** [comment]
I'm wondering if [this line](https://github.com/explosion/spaCy/blob/06324e5a5eda1a8cfa6be34e06ddf78f331cb497/spacy/pipeline/entity_linker.py#L336) causes the above observation? i.e. if the generated candidates contains multiple entities with the same score, shuffling change the original order.

What's the point of shuffling the generated candidates before scoring them?!

**Comment by svlandeg** [comment, marked as answer]
I agree that you'd expect the same performance if you set `incl_context` to False, because the `spacy` code will basically only retain the candidate with the highest scoring prior probability:
```
                           scores = xp.asarray(
                                [c.prior_prob for c in candidates]
                            )
                            best_index = scores.argmax().item()
                            best_candidate = candidates[best_index]
                            final_kb_ids.append(best_candidate.entity_)
```
There are two main reasons I can think of that would make the scores diverge with sciSpacy:
- sciSpacy may be applying some kind of threshold to the prior probability, and might be predicting `NIL` (or equivalent) when the prior probability is too low
- for some mentions, there are candidates with equally high prior probability. In this case, the shuffling of the list might have some influence on the final index that is selected. In general though, the shuffling shouldn't do any harm, as the list of candidates is not assumed to be ordered in any way (and so shuffling reduces potential bias)
